vm.go

```diff
@@ -617,6 +617,7 @@ func (vm *vm) run() {
 				ctx = context.Background()
 			}
 
+			vm.r.limiterWaitCount++
```

Comment:
we aren't actually waiting every time we call WaitN. To be able to determine if we are actually waiting take a look at [this](https://github.com/10gen/baas/blob/7ba0f60507d70ae82f3f85894840eb6996daebad/events/source.go#L1633-L1635) as an example. We'll only wait if the delay is > 0. 

[opt] in addition to the number of times we waited, I think it would also be valuable to track the total delay for the entire execution.

Reply:
Oh I see the distinction, does that come down to timing?

We have a rate limit of [10_000_000](https://github.com/10gen/baas/blob/6fc0339cb75c6614fc301a93fbf5e79e96021737/common/limits.go#L26) and a burst value of [250_000](https://github.com/10gen/baas/blob/6fc0339cb75c6614fc301a93fbf5e79e96021737/common/limits.go#L81) which I think is effectively 50_000 when you account for the [burst divisor](https://github.com/Calvinnix/goja/blob/d8c293c2ea8de8c64fa115b66a16931e7f942cb7/limiter.go#L16),

Aside from a lot of process yielding, I'm not sure I fully understand what scenario we actually wait and for how long.

If we use 10_000_000 ticks in half a second would we wait another half a second because that is the amount left until we can execute 50_000 ticks (provided by fillBucket)?

If we use 10_000_000 ticks with only 1ms left in the "second" window, would we only wait that 1ms before we can execute 50_000 ticks?

Reply:
If the above is true I assume we could do something like:

```
reservation := vm.r.limiter.Reserve(time.Now(), vm.r.limiterTicksLeft)
// count reservation.Delay()
// increment wait counter if Delay > 0
```

Reply:
```
			if r := vm.r.limiter.ReserveN(time.Now(), vm.r.limiterTicksLeft); r.OK() {
				waitDelayMS := r.Delay().Milliseconds()
				if waitDelayMS > 0 {
					vm.r.limiterWaitTotalMS += waitDelayMS
					vm.r.limiterWaitCount++
				}
			}
			if waitErr := vm.r.limiter.WaitN(ctx, vm.r.limiterTicksLeft); waitErr != nil {
```

Reply:
From the comment above `WaitN`:
> WaitN blocks until lim permits n events to happen.

In other words, we aren't going to wait if the limit allows those n events to be processed. To be able to determine if we're actually waiting or not, you want to do something along these lines:
```golang
	now = time.Now()
	r := vm.r.limiter.ReserveN(now, vm.r.limiterTicksLeft)
	if !r.OK() {
		panic("")
	}

	// Wait if necessary
	delay = r.DelayFrom(now)
	if delay > 0 {
		vm.r.limiterWaitCount++
		
		err := util.Sleep(ctx, delay)
		if err != nil {
			r.Cancel()
			panic(err)
		}
	}
```

> If we use 10_000_000 ticks in half a second would we wait another half a second because that is the amount left until we can execute 50_000 ticks (provided by fillBucket)?

Not exactly. We would approximately wait until enough time has passed to process the next 50k ticks. i.e. (50_000 / 10_000_000) => 0.005s => 5ms. You can find more about that [here](https://cs.opensource.google/go/x/time/+/refs/tags/v0.5.0:rate/rate.go;l=405).

> If we use 10_000_000 ticks with only 1ms left in the "second" window, would we only wait that 1ms before we can execute 50_000 ticks?

This can't happen. The rate limiter won't process more than 10_000_000 ticks / s. So it would just take 1s to process the first 10MM ticks. In other words, since we process 50k ticks at a time, a function can process at most 50k ticks every 5ms.

[These docs](https://pkg.go.dev/golang.org/x/time/rate#Limiter) do a decent job at explaining how this works. Let me know if you have more questions about it. I'm happy to also hop on a quick call.